### ISSUE TITLE
SUSI Linux and Installer separation

### ISSUE THREAD

**Issue by norbusan** [fossasia/susi_installer#1]
This issue discusses the following branches:
- susi_linux: [norbert/separate-installation](https://github.com/fossasia/susi_linux/tree/norbert/separate-installation)
- pi_gen: [norbert/separate-installer](https://github.com/fossasia/pi-gen/tree/norbert/separate-installer)
- susi_installer: [development](https://github.com/fossasia/susi_installer) (this branch here)

Ready-made RPi image
===================
A ready-made image built from the above PR is available at [SUSIbian_new-installer_20190509.img.xz](https://www.preining.info/susi.ai/SUSIbian_new-installer_20190509.img.xz)

For those too lazy to read all what is written here, I would appreciate testing of these image in various environments and tests, and let me know how it goes!

Aim of this group of changes
=======================
1. Separate the installer with the RPi specific parts into a separate repository
1. make `susi_linux` repository only containing the main application
1. restructure the layout of the `SUSI.AI`  folder on the RPi
1. allow for full factory reset and reset with keeping the settings `config.json`
1. unify installation between RPi and Desktop
1. support system wide installation on the Desktop
1. allow for future replacement of parts with Debian packages

Concerning the single points
======================

ad 1. Separate the installer with the RPi specific parts into a separate repository
------------------
The current `susi_linux` repository contains a lot of things besides the main application: all the installation logic and the RPi specific files. This makes separate development and packaging difficult. As soon as we start doing proper release management (in the future) I would like to see independent repositories so that independent releases can be done.

In the new `susi_installer` repo, I have separated the RPi specific stuff into the `raspi` sub-directory to make clear this is what is necessary for RPi, while desktop installations don't need that. Future additional hardware support might be added in the similar way.

ad 2. make `susi_linux` repository only containing the main application
-------------------------
See above concerning release management.

Further changes have been done that it is not necessary to have the CWD set to the susi_linux folder, but any folder will work, as long as it contains a `config.json`. This allows for multi-user support on the desktop

ad 3. restructure the layout of the `SUSI.AI`  folder on the RPi
---------------
The new layout of the `SUSI.AI` folder on the RPi is as follows:
```
    SUSI.AI/
        config.json
        susi_installer/
        susi_linux/
        susi_api_wrapper/
        susi_server/
        susi_skill_data/
        seeed-voicecard/
```
Here one important point is the fact that `config.json` is outside the directories. The layout is similar to what is described in the `developer_setup.md`.

ad 4. allow for full factory reset and reset with keeping the settings `config.json`
-------------
Following the above layout, a full reset would simply trash everything, while a reset with keeping config.json would only replace the sub-dirs. (Remark: Of course this can be done already now in the current layout)

ad 5. unify installation between RPi and Desktop
--------------
The new installer works the same way on the Desktop, that is, a new user or developer should be able to do
```
cd /tmp
wget https://raw.githubusercontent.com/fossasia/susi_installer/development/install.sh
bash install.sh
```
and get a working SUSI.AI.

ad 6. support system wide installation on the Desktop
--------------------
It is possible to install all components only once on a multi-user system and have multiple users use the same code but with separate configuration files for susi_linux. This is supported now with the new installer option `--system`, see the explanations in `--help`.

ad 7. allow for future replacement of parts with Debian packages
-----------------
In the future I want to build Debian packages `.deb` for `susi-server`, `susi-linux`, `susi-python`, `susi-skill-data` and simplify the installation to `apt-get` calls and the RPi specific stuff.

Questions, Remarks, and Current Status
===================
There are a few questions I have asked on Gitter or elsewhere, but I want to put here, too:

- (updated 20190424) At the moment we start susi-linux only after wap setup, to make sure that
  we have internet access. It would be nice to see susi-linux automatically being started when getting
  online, either via ethernet or via wap. (future work)

- I have tested that the generated RPi image (see above) boots on my device, automatically sets up `susi_linux` and answers to my questions. Also tested is the setup via the web interface of the WLAN system. **Not** tested is factory rest and other features.

- Sound play issues: It seems to me that after `Hi SUSI, play some music` (btw, SUSI, is always playing the same stupid Korean song :-( ) the sound output hangs, and the `vlc` process does not terminate. Not sure here - but I believe that is independent of the current discussion.

-----

Looking forward to an interesting discussion and lots of suggestions.

**Comment by norbusan** [comment]
Updates 20190414
- desktop installation is working from `desktop` branch, please test (see above and gitter forum)
- susi_linux is only started after wifi is set up, as in previous images

**Comment by norbusan** [comment]
Updates 20190426
- all the outstanding PR have been merged
- I have tested that installation on RPi with the new installer (that combines Desktop and RPi installer) works
- I have tested that installation on Desktop with the new installer works

**Comment by norbusan** [comment]
Updates 20190426-2
I have updated the RPi image uploaded to my server. Please test.

**Comment by norbusan** [comment]
Updates 20190509
I consider this now ready for prime time / feature complete. I have tested that the new installer works in the following circumstances:
- single user installation on the desktop 
- multi user installatio on the desktop (`install --system`)
- raspi installation via pi_gen

New RPi image (as listed above) are currently being uploaded.

I will create PR for the above branches.

**Comment by mariobehling** [comment]
Cool!

**Comment by norbusan** [comment]
Thanks @mariobehling for the approvals. As mentioned, I want to wait for @Orbiter to give his ok ...

**Comment by norbusan** [comment]
Merged by @Orbiter , so closing this issue
